### DISCUSSION TITLE
Need help regarding creating a service for a sub collection in MongoDB in Feathers v5

### DISCUSSION THREAD

**Discussion by gypsicoder** [feathersjs/feathers#3301] [Q&A]
Suppose I have a collection named "post" in Mongodb for which I have created a service named "post". And each "post" item has a property named "comments" which is an array containing "comment" objects. Now, I want to create a separate service for "comments". Is it possible in Feathers v5?
For example the API URL might be something like this  post/[postId]/comments or comments?postId=[postId]
I couldn't find any example for this type of problem. Someone please help!!!

**Comment by daffl** [comment]
Both is possible with a custom service. You can register it on the `post/:postId/comments` path and query `this.app.service('post')` for `params.route.postId` or on `/comments` and use `params.query.postId`. Then you'd implement the [service methods](https://feathersjs.com/api/services.html) to manipulate the comments for each post accordingly.

**Comment by gypsicoder** [comment]
@daffl 
Thank you for your guidelines. I have made the service "post" with mongodb collection named "post" and set that model. Now, I am confused how to set the model for "comments"? And if I write a custom service how to do the `pagination`, `sort` and `select` stuffs?
Can you please show me some example with writing custom service for this type of scenario? I didn't find any in the FeathersJS website.

**Reply by daffl** [reply]
To be honest, I'm not sure if it's a good idea to store the comments in an array. I'd make a separate collection where you can store them and relate them via the `postId`. The [chat guide](https://feathersjs.com/guides/basics/generator.html) shows how to relate and populate entities with MongoDB.

**Reply by gypsicoder** [reply]
It's a requirement to do in this way. I had been successful to create the service using separate collection which is very straight forward. But, I need the solution for the given scenario.

**Reply by daffl** [reply]
Have you tried generating a custom service with what I suggested? It scaffolds a lot of things for you and then you need to implement the logic you need for each method. For example to return all comments for a post in `find` it would be something like this:

```ts
export class CommentsService<ServiceParams extends CommentsParams = CommentsParams>
  implements ServiceInterface<Comments, CommentsData, ServiceParams, CommentsPatch>
{
  constructor(public options: CommentsServiceOptions) {}

  async find(params: ServiceParams = {}): Promise<Comments[]> {
    // via `?postId=<postId>` query parameter
    const { postId } = params.query || {}
    // or on the `post/:postId/comment route
    const { postId } = params.route || {}
    
    const { comments } = await this.options.app.service('posts').get(postId)

    return comments
  }

  async get(id: Id, _params?: ServiceParams): Promise<Comments> {
    return {
      id: 0,
      text: `A new message with ID: ${id}!`
    }
  }

  async create(data: CommentsData, params?: ServiceParams): Promise<Comments>
  async create(data: CommentsData[], params?: ServiceParams): Promise<Comments[]>
  async create(data: CommentsData | CommentsData[], params?: ServiceParams): Promise<Comments | Comments[]> {
    if (Array.isArray(data)) {
      return Promise.all(data.map((current) => this.create(current, params)))
    }

    return {
      id: 0,
      ...data
    }
  }

  // This method has to be added to the 'methods' option to make it available to clients
  async update(id: NullableId, data: CommentsData, _params?: ServiceParams): Promise<Comments> {
    return {
      id: 0,
      ...data
    }
  }

  async patch(id: NullableId, data: CommentsPatch, _params?: ServiceParams): Promise<Comments> {
    return {
      id: 0,
      text: `Fallback for ${id}`,
      ...data
    }
  }

  async remove(id: NullableId, _params?: ServiceParams): Promise<Comments> {
    return {
      id: 0,
      text: 'removed'
    }
  }
}
```

**Reply by gypsicoder** [reply]
Thank you for this. I have came to this point. But, I don't know how to do the `sort`, `pagination` for the `comments` here. Would you can show that also?

**Reply by daffl** [reply]
MongoDB does not have an option to do sorting and pagination with embedded arrays. You will have to do that manually yourself (e.g. by slicing the array based on the `$sort`, `$limit` and `$skip` parameters).

**Reply by gypsicoder** [reply]
@daffl 
I have found a solution to do sort, and pagination using `slice` on mongo using [Aggregate Framework](https://www.mongodb.com/docs/manual/core/aggregation-pipeline/)
